modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java

```diff
@@ -144,8 +156,17 @@ static ExecutionTarget colocate(OneOfTarget oneOf, PartitionedTarget partitioned
             throw new ColocationMappingException("Targets are not colocated");
         }
 
-        if (!isPow2((partitioned.nodes))) {
-            throw new ColocationMappingException("Targets are not colocated");
+        if (isPow2((partitioned.nodes))) {
+            return partitioned; // All partitions on single node.
+        }
+
+        long colocatedNodes = oneOf.nodes;
+        for (int partNo = 0; partNo < partitioned.partitionsNodes.length; partNo++) {
+            colocatedNodes &= partitioned.partitionsNodes[partNo];
+
+            if (colocatedNodes == 0) {
+                throw new ColocationMappingException("Targets are not colocated");
+            }
         }
 
         return partitioned;
```

Comment:
you have to create new partitioned target with `colocatedNodes` for every partition. Here is an example: result of colocation of `PartitionedTarget(1=[A, B, C], 2=[C, D, E])` and `OneOfTarget(C, F)` is `PartitionedTarget(1=[C], 2=[C])`